### DISCUSSION TITLE
Cannot Find Module

### DISCUSSION THREAD

**Discussion by johnwargo** [microsoft/vscode-discussions#1100] [Extension Development QnA] [answered]
I published my first Visual Studio Code extension recently, but even though the extension works properly in Visual Studio Code while debugging the extension, it doesn't work when installed from the Marketplace or when loading the `.vsix` file manually in Visual Studio Code. 

The extension is here: https://marketplace.visualstudio.com/items?itemName=john-wargo.describer-genai
The source code is here:  https://github.com/johnwargo/vscode-describer-genai

I wasn't sure how to debug this issue, but the folks at Stack Overflow pointed me at the debugging tools: https://stackoverflow.com/questions/78173117/visual-studio-code-extension-commands-not-registering

When I load the extension via the `.vsix` file and try to invoke my plugin's settings command, the console gets the following error:

```
mainThreadExtensionService.ts:113 Activating extension 'john-wargo.describer-genai' failed: Cannot find module 'openai'
Require stack:
- c:\Users\john\.vscode\extensions\john-wargo.describer-genai-0.1.1\out\extension.js
- c:\Users\john\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\loader.js
- c:\Users\john\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-amd.js
- c:\Users\john\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js
- .
```

I double checked that the dependency is properly defined in the `package.json` file's `dependencies` section, I also deleted the `node_modules` folder then ran `npm i` in the extension project folder. 

Can someone please help me resolve this issue? From everything I know about node and npm this should work. I assume that when Code installs the extension it installs dependencies automatically.

**Comment by tamuratak** [comment, marked as answer]
That's because your `.vscodeignore` file includes the following line, which tell  the `vsce` command to excludes  the `node_modules` folder from the `.vsix` file.

```
node_modules/**
```

https://github.com/johnwargo/vscode-describer-genai/blob/eb67fbe573b903716f350fc2d7abb5393d37abbc/.vscodeignore#L12

**Reply by johnwargo** [reply]
Aaah, thank you!  I only put that there because I saw it in an article somewhere. I'll try that out.

**Reply by johnwargo** [reply]
that was it, thank you!

**Reply by johnwargo** [reply]
@tamuratak I just looked at the source code for a few popular VSCode extensions in marketplace and they all had that line (or variants thereof) in their `.vscodeignore` files. 

Examples:

* https://github.com/microsoft/vscode-jupyter/blob/main/.vscodeignore
* https://github.com/microsoft/vscode-edge-devtools/blob/main/.vscodeignore

So I'm curious why that was an issue for me.

**Reply by tamuratak** [reply]
Because, with webpack or esbuild, they are combining JavaScript files under the `node_modules` folder into a single big file and are shipping their extension with the big file, they don't have to include  the `node_modules` folder in the `.vsix` file.  It is called ["bundling"](https://dev.to/how-to-dev/why-do-we-use-javascript-bundlers-3e60).

- https://github.com/microsoft/vscode-edge-devtools/blob/565179e8b27afae7e75664d38e272ab4860a8836/package.json#L668
- https://github.com/microsoft/vscode-jupyter/blob/076cb1cbcc46779c3d37b9205ea951e3862b9e48/package.json#L2126


See the document. Bundling is recommended, but not necessary. 

- https://code.visualstudio.com/api/working-with-extensions/bundling-extension

**Reply by johnwargo** [reply]
Thanks @tamuratak. Yes, I understand what bundling is.  I assumed that since extensions used a `package.json` file that VSCode installed dependencies during extension installation. I wasn't expecting that all dependencies would be in the package. Thanks for your help.
